File: app/src/androidTest/java/com/backbase/citysearch/SearchAlgoTest.java
package com.backbase.citysearch;

import android.content.Context;

import androidx.test.platform.app.InstrumentationRegistry;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.backbase.citysearch.datastorage.CityDataStore;
import com.backbase.citysearch.datastorage.CityDataStoreImpl;
import com.backbase.citysearch.models.City;
import com.backbase.citysearch.utilities.DataUtil;

import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.List;

import static org.junit.Assert.*;

/**
 * Instrumented test, which will execute on an Android device.
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
@RunWith(AndroidJUnit4.class)
public class SearchAlgoTest {
    @Test
    public void searchWithWrongEntryShouldReturnNull() {
        List<City> matchingCities = getCityDataStore().findMatchingCities("garbage");
        assertNull(matchingCities);
    }

    @Test
    public void searchWithEmptyEntryShouldReturnAllItems() {
        List<City> matchingCities = getCityDataStore().findMatchingCities("");
        assertNotNull(matchingCities);
        assertEquals(209557, matchingCities.size());
    }

    @Test
    public void searchShouldReturnCorrectItems() {
        assertEquals(4, getCityDataStore().findMatchingCities("Sydney").size());
    }

    private CityDataStore getCityDataStore() {
        CityDataStore cityDataStore = new CityDataStoreImpl();
        Context appContext = InstrumentationRegistry.getInstrumentation().getTargetContext();
        DataUtil.addCitiesToDB(appContext,cityDataStore);
        return cityDataStore;
    }
}

File: app/src/androidTest/java/com/backbase/citysearch/SearchAlgoTest.kt
package com.backbase.citysearch

import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import com.backbase.citysearch.datastorage.CityDataStore
import com.backbase.citysearch.datastorage.CityDataStoreImpl
import com.backbase.citysearch.utilities.DataUtil.addCitiesToDB
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith

/**
 * Instrumented test, which will execute on an Android device.
 *
 * @see [Testing documentation](http://d.android.com/tools/testing)
 */
@RunWith(AndroidJUnit4::class)
class SearchAlgoTest {
    @Test
    fun searchWithWrongEntryShouldReturnNull() {
        val matchingCities = cityDataStore.findMatchingCities("garbage")
        Assert.assertNull(matchingCities)
    }

    @Test
    fun searchWithEmptyEntryShouldReturnAllItems() {
        val matchingCities = cityDataStore.findMatchingCities("")
        Assert.assertNotNull(matchingCities)
        Assert.assertEquals(209557, matchingCities!!.size.toLong())
    }

    @Test
    fun searchShouldReturnCorrectItems() {
        Assert.assertEquals(4, cityDataStore.findMatchingCities("Sydney")!!.size.toLong())
    }

    private val cityDataStore: CityDataStore
        private get() {
            val cityDataStore: CityDataStore = CityDataStoreImpl()
            val appContext = InstrumentationRegistry.getInstrumentation().targetContext
            addCitiesToDB(appContext, cityDataStore)
            return cityDataStore
        }
}